subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/dsl/DefaultComponentMetadataHandler.java

```diff
@@ -131,7 +135,7 @@ public ModuleComponentResolveMetadata processMetadata(ModuleComponentResolveMeta
             updatedMetadata = metadata;
         } else {
             MutableModuleComponentResolveMetadata mutableMetadata = metadata.asMutable();
-            ComponentMetadataDetails details = instantiator.newInstance(ComponentMetadataDetailsAdapter.class, mutableMetadata);
+            ComponentMetadataDetails details = instantiator.newInstance(ComponentMetadataDetailsAdapter.class, mutableMetadata, instantiator, dependencyMetadataNotationParser);
             processAllRules(metadata, details);
             updatedMetadata = mutableMetadata.asImmutable();
```

Comment:
⭕️ asImmutable now needs to take into account that we might have changed the dependencies, so it needs to make a defensive copy if we did.